relayer-engine/src/storage/storage.ts

```diff
@@ -209,6 +216,84 @@ export class Storage {
     );
   }
 
+  async getReadyWorkflows(start: number, end: number): Promise<Workflow[]> {
+    return this.getWorkflowsInQueue(
+      this.constants.READY_WORKFLOW_QUEUE,
+      start,
+      end,
+    );
+  }
+
+  async getFailedWorkflows(start: number, end: number): Promise<Workflow[]> {
+    return this.getWorkflowsInQueue(
+      this.constants.DEAD_LETTER_QUEUE,
+      start,
+      end,
+    );
+  }
+
+  async getInProgressWorkflows(
+    start: number,
+    end: number,
+  ): Promise<Workflow[]> {
+    return this.getWorkflowsInQueue(
+      this.constants.ACTIVE_WORKFLOWS_QUEUE,
+      start,
+      end,
+    );
+  }
+
+  async getDelayedWorkflows(start: number, end: number): Promise<Workflow[]> {
+    return this.getWorkflowsInQueue(
+      this.constants.DELAYED_WORKFLOWS_QUEUE,
+      start,
+      end,
+    );
+  }
+
+  async moveFailedWorkflowToReady(workflowId: {
+    pluginName: string;
+    id: string;
+  }) {
+    const workflowKey = this._workflowKey(workflowId);
+    return this.store.withRedis(async redis => {
+      const [removedFromDelayed, removedFromDeadLetter] = await Promise.all([
+        redis.lRem(this.constants.DELAYED_WORKFLOWS_QUEUE, 0, workflowKey),
+        redis.lRem(this.constants.DEAD_LETTER_QUEUE, 0, workflowKey),
+      ]);
+      if (!removedFromDeadLetter && !removedFromDelayed) {
+        throw new Error("workflow wasn't failed or retrying");
+      }
+      await redis.hSet(workflowKey, <SerializedWorkflowKeys>{
+        failedAt: "",
+        processingBy: "",
+        errorMessage: "",
+        errorStacktrace: "",
+        startedProcessingAt: "",
+        retryCount: 0,
+        completedAt: "",
+      });
+      await redis.lPush(this.constants.READY_WORKFLOW_QUEUE, workflowKey);
+      const raw = await redis.hGetAll(workflowKey);
```

Comment:
Not actionable, but seeing all these separate redis calls is making me nervous about getting into inconsistent state. I don't think it's happened yet, but lack of transactionality will bite us eventually

Reply:
I'll try to tackle this in a separate pr. I think we can do it in a multi tx but we may need to change our types a little bit? Btw, since we're only using redis, maybe we can just use the redis type directly instead of `IRedis`?

Reply:
Sounds good on the multi 👍 

As far as IRedis, the main reason I kept that around was it has much simpler types than the real interface, which I found hard to read/grok and we initially did not use much of the redis api. At this point I could go either way on keeping it.